text/0006-monolothic-packaging.md

```diff
@@ -0,0 +1,343 @@
+- [ ] `constructs`
+- [ ] `@aws-cdk/assert` library. Do we have to do the assert library rewrite as a jsii module (we want to do that anyway!).
+- [ ] Rewrite all example code (see [cdk-rewrite-mono-import](https://github.com/rix0rrr/cdk-rewrite-mono-imports)).
+- [ ] Reference documentation needs to also support submodules/namespaces and use the submodule's README file.
+
+# Drawbacks
+
+- [ ] In JavaScript/TypeScript code will always be fully qualified
+  (`aws_s3.Bucket` instead of `Bucket`) due to limitations of the `import` tool.
+- [ ] The size of the single module will be large (~50MiB) which can create issues of using the CDK in non-standard environments such as from a Lambda function.
+- [ ] Breaking change
+
+# Rationale and Alternatives
+
+I believe our users think of the AWS CDK as a "standard library" (or a
+"framework") and not as another library that they depend on to build their
+applications (like the AWS SDK for example). When users create CDK libraries and
+apps they don't think of the AWS CDK as yet another library that they use, they
+think of the AWS CDK as the foundation of their app.
+
+We can draw the analogy to other standard libraries like the Node.js SDK, the
+JDK, the .NET Framework. When users write Node.js libraries or apps they don't "depend" on the Node.js SDK, they actually "peer depend" on it (it is expressed via a special `engines` section in `package.json` but for all intents and purposes it's behaves like a peer dependency).
+
+I would argue that if a vendor publishes a 3rd-party construct library, what they *really* want to say is "I am compatible with CDK >= 1.23.0". Then, the decision about which actual CDK version is being used is left to the app level.
+
+## Alternatives Considered
+
+### Tooling
+
+TODO: we considered building additional tooling (e.g. `cdk install`) that will make hyper-modular peer dependencies sane.
+
+### Meta-package
+
+This approach proposes that we keep the hyper modularity but also ship a meta-package that will either just take a dependency on all modules (like `decdk` is doing today) or also "re-export" all the types.
+
+The main benefit is that it will technically allow the interoperability of the
+two modes. Libraries can still take granular peer dependencies while apps will
+depend on the meta package.
+
+TODO: we need to check if in JavaScript it is possible to avoid the re-export and only create a module that brings in all the other modules as dependencies. The answer is probably "depends on the package manager". Yarn pre-2.0 for example would hoist all the transitive deps so it would "just work", but post 2.0 they actually disallow that.
+
+# Adoption Strategy
+
+TODO
+
+# Unresolved questions
+
+TODO
+
+# Future Possibilities
+
+TODO
+
+# Appendix
+
+## Peer Dependencies in npm
+
+> Based on a whitepaper by [@rix0rrr](https://github.com/rix0rrr).
+
+Most languages (Java, Python, .NET) only support a single instance of a module
```

Comment:
Might want to mention this can become an issue in Python as well unless extra care is taken, since Python does not reliably resolve peer dependencies to an expected version by default.

Reply:
Removed this section.